src/core/importType.js

```diff
@@ -87,17 +74,48 @@ function isRelativeToSibling(name) {
   return /^\.[\\/]/.test(name);
 }
 
-function typeTest(name, context, path) {
+function isExternalPath(path, context) {
+  if (!path) {
+    return false;
+  }
+
   const { settings } = context;
+  const packagePath = getContextPackagePath(context);
+
+  if (relative(packagePath, path).startsWith('..')) {
+    return true;
+  }
+
+  const folders = (settings && settings['import/external-module-folders']) || ['node_modules'];
+  return folders.some((folder) => {
+    const folderPath = nodeResolve(packagePath, folder);
+    const relativePath = relative(folderPath, path);
+    return !relativePath.startsWith('..');
+  });
+}
+
+function isInternalPath(path, context) {
+  if (!path) {
+    return false;
+  }
+  const packagePath = getContextPackagePath(context);
+  return !relative(packagePath, path).startsWith('../');
+}
+
+function isExternalLookingName(name) {
+  return isModule(name) || isScoped(name);
+}
+
+function typeTest(name, context, path, settings = context.settings) {
```

Comment:
can you remind me, when do we want to pass a `settings` that is different than `context.settings`? iow, why are we adding this fourth argument?

Reply:
Good question.

My original reasoning was that since `isExternalModule` / `isExternalModuleMain` are being exported, they should be considered part of the public API (though this may be incorrect—I don't know—it seemed like a safer assumption than the opposite conclusion). And based on how they're working in main, it seems in theory possible for a consumer to pass in separate `settings` which would take precedence over `context.setting`. I wasn't sure whether that use case was important or not, but figured that any behavior change could be a separate concern for a different PR. Internally to `eslint-plugin-import`, the only source calls to `isExternalModule` that I'm finding all provide `context.settings` as settings—so, no internal use cases for `settings` being different than `context.settings`.

Anyways, because of the potential for external consumers relying on this "feature", I figured I should keep the behavior constant. At the time, the simplest way to me seemed to be to provide settings as an argument to `typeTest` as well, but alternatively we could do something like `typeTest(name, {...context, settings}, path)` from `isExternalModule`. I thought that'd be messier, but now that I'm re-thinking it seems reasonably simple, and probably a lot better than adding unnecessary complexity to `typeTest`. I'll update to that.

 I would be in favor of a change that removed the `settings` argument entirely from `isExternalModule` and `isExternalModuleMain`. If consumers do need to override `settings` for some reason, they could migrate from `isExternalModule(name, settings, path, context)` => `isExternalModule(name, {...context, settings}, path)`. My default inclination is to do that in a separate PR, but I'd be perfectly happy to include the change into this PR as well.

Reply:
OK, so I just tried to do `typeTest(name, { ...context, settings }, path)` but that fails because the eslint context object has a non standard prototype, and so for example `getFilename` is now missing from the attempted copy (producing errors). With a more faithful copy, so `typeTest(name, Object.assign(Object.create(Object.getPrototypeOf(context)), context, { settings }, path)`, I'm now seeing:

```
TypeError: Cannot assign to read only property 'settings'
```

Which is probably fixable with some more work, but the solution's complexity is already making me think that this might not be the ideal path to try and follow. Any thoughts?

Reply:
If there's no internal use cases for passing a settings that's different than `context.settings`, then it's better not to permit that.